# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/renderer/mod.rs
# Commit: 6d98d185a076
# Full Hash: 6d98d185a07684b3538dd0052cb14b67cba38568
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2021-11-05 09:34:21
# Regressor Bug: 1730695
# File Overlap Count: 1
# Description:
#   Bug 1730695 - Fix panic when casting large value to i32 rect r=gfx-reviewers,bradwerth,nical,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D128122
# ==============================================================================

diff -r bba94c79f3e1 -r 6d98d185a076 gfx/wr/webrender/src/renderer/mod.rs
--- a/gfx/wr/webrender/src/renderer/mod.rs	Thu Nov 04 19:20:21 2021 +0000
+++ b/gfx/wr/webrender/src/renderer/mod.rs	Thu Nov 04 19:44:32 2021 +0000
@@ -4404,6 +4404,7 @@
 
             if can_use_partial_present {
                 let mut combined_dirty_rect = DeviceRect::zero();
+                let fb_rect = DeviceRect::from_size(draw_target_dimensions.to_f32());
 
                 // Work out how many dirty rects WR produced, and if that's more than
                 // what the device supports.
@@ -4415,7 +4416,14 @@
                         &tile.local_dirty_rect,
                         tile.transform_index,
                     );
-                    combined_dirty_rect = combined_dirty_rect.union(&dirty_rect);
+
+                    // In pathological cases where a tile is extremely zoomed, it
+                    // may end up with device coords outside the range of an i32,
+                    // so clamp it to the frame buffer rect here, before it gets
+                    // casted to an i32 rect below.
+                    if let Some(dirty_rect) = dirty_rect.intersection(&fb_rect) {
+                        combined_dirty_rect = combined_dirty_rect.union(&dirty_rect);
+                    }
                 }
 
                 let combined_dirty_rect = combined_dirty_rect.round();
